src/common/column_matrix.h

```diff
@@ -103,96 +110,179 @@ class ColumnMatrix {
+    }
 
-    // store least bin id for each feature
-    index_base_.resize(nfeature);
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      index_base_[fid] = gmat.cut.Ptrs()[fid];
+    index_.resize(boundary_[nfeature - 1].index_end * type_size_, 0);
+    if (!all_dense) {
+      row_ind_.resize(boundary_[nfeature - 1].index_end);
     }
 
+    // store least bin id for each feature
+    index_base_ = const_cast<uint32_t*>(gmat.cut.Ptrs().data());
+
     // pre-fill index_ for dense columns
+    const bool noMissingValues = gmat.row_ptr[nrow] == nrow * nfeature;
 
-    #pragma omp parallel for
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      if (type_[fid] == kDenseColumn) {
-        const size_t ibegin = boundary_[fid].index_begin;
-        uint32_t* begin = &index_[ibegin];
-        uint32_t* end = begin + nrow;
-        std::fill(begin, end, std::numeric_limits<uint32_t>::max());
-        // max() indicates missing values
+    if (noMissingValues) {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, false);
+    } else {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, true);
+    }
+
+    if (all_dense) {
+      switch (gmat.index.getBinBound()) {
+        case POWER_OF_TWO_8:
+          SetIndexAllDense(gmat.index.data<uint8_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_16:
+          SetIndexAllDense(gmat.index.data<uint16_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_32:
+          SetIndexAllDense(gmat.index.data<uint32_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+      }
+    } else {
+      switch (type_size_) {
+        case 1:
+          SetIndex<uint8_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 2:
+          SetIndex<uint16_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 4:
+          SetIndex<uint32_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+      }
+    }
+  }
+
+  /* Fetch an individual column. This code should be used with XGBOOST_TYPE_SWITCH
+     to determine type of bin id's */
+  template <typename T>
+  inline Column<T> GetColumn(unsigned fid) const {
+    Column<T> c(type_[fid],
+                reinterpret_cast<const T*>(&index_[boundary_[fid].index_begin * type_size_]),
+                index_base_[fid], (type_[fid] == ColumnType::kSparseColumn ?
+                &row_ind_[boundary_[fid].index_begin] : nullptr),
+                boundary_[fid].index_end - boundary_[fid].index_begin,
+                &missing_flags_, boundary_[fid].index_begin);
+    return c;
+  }
+
+  template<typename T>
+  inline void SetIndexAllDense(T* index, const GHistIndexMatrix& gmat,  const size_t nrow,
+                               const size_t nfeature,  const bool noMissingValues) {
+    T* local_index = reinterpret_cast<T*>(&index_[0]);
+
+    if (noMissingValues) {
```

Comment:
It's confusing that the function is called `AllDense` but still have missing values.  Could you please refactor the code to clarify the logic a bit?

Reply:
As far as I understand missing values are only possible for Dense columns (and it's even handled only in PartitionDenseKernel). For sparse column it doesn't make sense.
The logic are still the same as were before: by _default_ all elements in Sparse/Dense column are interpreted as missing values - max `uint32` value for old version, and corresponding `true` missing flag for current changes).
But if we detect that there is no missing value:
`gmat.row_ptr[nrow] == nrow * nfeature;`
we can set all flags to `false` values and apply threading for `index_` setting.

Reply:
added.